File: Automation/001-AutomationFramework/C#/AutomationFramework/UIMappings/Pages/EmailProviders/Mailinator.cs
using CommonCore.Tests;
using NUnit.Framework;
using OpenQA.Selenium;
using SeleniumEngine.Base;
using SeleniumEngine.DriverInitialization;
using SeleniumEngine.Extensions;

namespace UIMappings.Pages.EmailProviders
{
    public class Mailinator : BasePage
    {
        #region Fields and Properties
        public IWebElement Email => WebDriverFactory.WebDriver.FindElement(By.XPath("//input[@id='search'] | //input[@id='inbox_field']"));
        public IWebElement Go => WebDriverFactory.WebDriver.FindElement(By.XPath("//button[normalize-space(text())='GO']"));
        public IWebElement EmailListContainer => WebDriverFactory.WebDriver.FindElement(By.XPath("//*[@class='table-striped jambo_table']"));
        public IWebElement EmailContent => WebDriverFactory.WebDriver.FindElement(By.TagName("body"));
        public IWebElement EmailName => WebDriverFactory.WebDriver.FindElement(By.XPath("//*[@id='inbox_field']"));
        public IWebElement DeleteButton => WebDriverFactory.WebDriver.FindElement(By.XPath("//button[@aria-label='Delete Button']"));
        public IWebElement NotificationEmailContent => WebDriverFactory.WebDriver.FindElement(By.XPath("//*[@id='secureLinkSalutationId']//..//..//p[1]"));
        #endregion

        #region Methods
        public Mailinator EnterEmail(string emailAddress)
        {
            this.Email.SendKeysEx(emailAddress, "EmailAddress");
            return this;
        }

        public Mailinator ClickOnGo()
        {
            this.Go.ClickEx("Go");
            return this;
        }

        public Mailinator SaveGeneratedEmailAddress(string email)
        {
            TestInMemoryParameters.Instance.EmailAddress = email;
            return this;
        }

        public string GetGeneratedEmail()
        {
            return string.Format("{0}@{1}", this.EmailName.GetValueEx(), "mailinator.com");
        }

        public Mailinator ClickOnEmail(string subject)
        {
            this.EmailListContainer.FindElements(By.XPath("./tbody/tr"))
                              .FirstOrDefault(x => x.FindElementEx(By.XPath("./td[3]")).GetValueEx().Contains(subject))
                              .ClickEx(subject);
            return this;
        }

        private bool IsEmailReceived(string subject)
        {
            IWebElement emailElement = this.EmailListContainer.FindElements(By.XPath("./tbody/tr"))
                                                         .FirstOrDefault(x => x.FindElementEx(By.XPath("./td[3]")).GetValueEx().Contains(subject));

            return emailElement != null;
        }

        public Mailinator GetVerificationCode(string text, out string verificationCodeText)
        {
            WebDriverFactory.WebDriver.SwitchTo().Frame("html_msg_body");

            IWebElement textElement = WebDriverFactory.WebDriver.FindElementEx(By.XPath(string.Format("//*[contains(text(), '{0}')]", text)));

            if (textElement == null)
            {
                throw new NotFoundException(string.Format("Failed to locate '{0}' in the content", text));
            }

            verificationCodeText = textElement.Text;

            return this;
        }

        public Mailinator WaitForEmail(string subject)
        {
            uint timeoutInSeconds = 240;
            int pollingIntervalInSeconds = 15;

            try
            {
                WebDriverFactory.WebDriver.WaitUntil(_ => IsEmailReceived(subject), timeoutInSeconds: timeoutInSeconds, pollingIntervalInSeconds: pollingIntervalInSeconds);
            }
            catch (Exception ex)
            {
                throw new TimeoutException(string.Format("Email not received after {0} seconds", timeoutInSeconds), ex);
            }

            return this;
        }

        public Mailinator VerifyNumberOfEmailsWithSpecificSubject(string subjectEmail, string expectedNumberOfEmails)
        {
            string emailXpath = string.Format("//div[@id='inboxpane']//a[contains(text(), '{0}')]", subjectEmail);
            var emailElements = WebDriverFactory.WebDriver.FindElementsEx(By.XPath(emailXpath), 10);

            Assert.AreEqual(expectedNumberOfEmails, emailElements.Count.ToString(), string.Format("Expected number of {0} emails not not matching the actual number {1}"), expectedNumberOfEmails, emailElements.Count);

            WebDriverFactory.TakeScreenshot("Mailinator MAIL - Verify number of emails");

            return this;
        }

        public Mailinator SelectAllEmailsFromInbox()
        {
            var emailItems = WebDriverFactory.WebDriver.FindElements(By.XPath("//*[@type='checkbox']"));

            foreach (var item in emailItems)
                item.ClickEx("emailItem");

            return this;
        }

        public Mailinator ClickOnDelete()
        {
            this.DeleteButton.ClickEx("Delete");
            return this;
        }

        public Mailinator VerifyEmailBodyContent(string text)
        {
            WebDriverFactory.WebDriver.SwitchTo().Frame("html_msg_body");
            string email = this.NotificationEmailContent.GetValueEx();
            Assert.IsTrue(email.Contains(text), string.Format("Email body does not contain expected text! \nExpected: {0} \nActual: {1}", text, email));
            return this;
        }
        #endregion
    }
}

File: Automation/Woolsocks/TestAutomationFramework/src/test/java/uiMappings/pages/emailProviders/Mailinator.java
package uiMappings.pages.emailProviders;

import engines.selenium.base.BasePageActions;
import engines.selenium.driverInitialization.WebDriverFactory;
import org.junit.jupiter.api.Assertions;
import org.openqa.selenium.By;
import org.openqa.selenium.NotFoundException;
import org.openqa.selenium.TimeoutException;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;

import java.util.List;

public class Mailinator extends BasePageActions {
    // region Fields and Properties
    @FindBy(xpath = "//input[@id='search'] | //input[@id='inbox_field']")
    public WebElement email;
    @FindBy(xpath = "//button[normalize-space(text())='GO']")
    public WebElement go;
    @FindBy(xpath = "//*[@class='table-striped jambo_table']")
    public WebElement emailListContainer;
    @FindBy(tagName = "body")
    public WebElement emailContent;

    @FindBy(xpath = "//a[contains(@class, 'inbox-link')]")
    public WebElement publicInbox;

    @FindBy(xpath = "//*[@id='inbox_field']")
    public WebElement emailName;
    @FindBy(xpath = "//button[@aria-label='Delete Button']")
    public WebElement deleteButton;
    @FindBy(xpath = "//*[@id='secureLinkSalutationId']//..//..//p[1]")
    public WebElement notificationEmailContent;
    // endregion

    // region Methods
    public Mailinator() {
        PageFactory.initElements(WebDriverFactory.getThreadSafeInstance().driver, this);
    }

    public Mailinator enterEmail(String emailAddress) {
        sendKeysEx(this.email, emailAddress, "EmailAddress", false, false);
        return this;
    }

    public Mailinator clickOnGo() {
        clickEx(this.go, "Go", false);
        return this;
    }

    public Mailinator clickOnEmail(String subject) throws Exception {
        List<WebElement> rows = this.emailListContainer.findElements(By.xpath("./tbody/tr"));
        WebElement emailRow = rows.stream().map(x -> x.findElement(By.xpath("./td[3]"))).findFirst().get();

        String emailRowValue = getValueEx(emailRow);
        if (emailRowValue.contains(subject)) {
            clickEx(emailRow, subject, false);
        } else {
            throw new Exception(String.format("Failed to find the email containing subject: %s", subject));
        }
        return this;
    }

    private boolean isEmailReceived(String subject) {
        WebElement emailElement = this.emailListContainer.findElements(By.xpath("./tbody/tr")).stream().
                map(x -> x.findElement(By.xpath("./td[3]"))).findFirst().get();

        return getValueEx(emailElement).contains(subject);
    }

    public Mailinator waitForEmail(String subject) {
        int timeoutInSeconds = 240;

        try {
            WebDriverFactory.getThreadSafeInstance().wait.until($ -> isEmailReceived(subject));
        } catch (Exception ex) {
            throw new TimeoutException(String.format("Email not received after {0} seconds", timeoutInSeconds), ex);
        }

        return this;
    }

    public Mailinator verifyEmailBodyContent(String text) {
        WebDriverFactory.getThreadSafeInstance().driver.switchTo().frame("html_msg_body");
        String email = getValueEx(this.notificationEmailContent);
        Assertions.assertTrue(email.contains(text), String.format("Email body does not contain expected text! \nExpected: {0} \nActual: {1}", text, email));
        return this;
    }

    public String clickVerificationLink(String text) {
        WebDriverFactory.getThreadSafeInstance().driver.switchTo().frame("html_msg_body");

        WebElement textElement = WebDriverFactory.getThreadSafeInstance().driver.findElement(By.xpath(String.format("//*[contains(text(), '{0}')]", text)));

        if (textElement == null) {
            throw new NotFoundException(String.format("Failed to locate '{0}' in the content", text));
        }

        return textElement.getText();
    }

    public Mailinator clikOnPublicInbox() {
        clickEx(this.publicInbox, "PublicInbox", false);
        return this;
    }
    // endregion
}
